internal/pkg/cli/app_package.go

```diff
@@ -62,94 +61,122 @@ func NewPackageAppOpts() *PackageAppOpts {
+		if !contains(o.AppName, names) {
+			return fmt.Errorf("application '%s' does not exist in the workspace", o.AppName)
 		}
-		app, err := opts.prompt.SelectOne(appPackageAppNamePrompt, "", names)
-		if err != nil {
-			return fmt.Errorf("prompt application name: %w", err)
-		}
-		opts.AppName = app
 	}
-	if opts.EnvName == "" {
-		names, err := opts.listEnvNames()
-		if err != nil {
+	if o.EnvName != "" {
+		if _, err := o.store.GetEnvironment(o.ProjectName(), o.EnvName); err != nil {
 			return err
 		}
-		if len(names) == 0 {
-			return fmt.Errorf("there are no environments in project %s", opts.ProjectName())
-		}
-		env, err := opts.prompt.SelectOne(appPackageEnvNamePrompt, "", names)
-		if err != nil {
-			return fmt.Errorf("prompt environment name: %w", err)
-		}
-		opts.EnvName = env
 	}
 	return nil
 }
 
-// Validate returns an error if the values provided by the user are invalid.
-func (opts *PackageAppOpts) Validate() error {
-	if opts.ProjectName() == "" {
-		return errNoProjectInWorkspace
-	}
-	if opts.Tag == "" {
-		tag, err := getVersionTag(opts.runner)
-		if err != nil {
-			return fmt.Errorf("image tag cannot be empty, please provide the %s flag", color.HighlightCode("--tag"))
-		}
-		opts.Tag = tag
-	}
-	if opts.AppName != "" {
-		names, err := opts.listAppNames()
-		if err != nil {
-			return err
-		}
-		if !contains(opts.AppName, names) {
-			return fmt.Errorf("application '%s' does not exist in the workspace", opts.AppName)
-		}
+// Ask prompts the user for any missing required fields.
+func (o *PackageAppOpts) Ask() error {
+	if err := o.askAppName(); err != nil {
+		return err
 	}
-	if opts.EnvName != "" {
-		if _, err := opts.store.GetEnvironment(opts.ProjectName(), opts.EnvName); err != nil {
-			return err
-		}
+	if err := o.askEnvName(); err != nil {
+		return err
 	}
-	return nil
+	return o.askTag()
 }
 
 // Execute prints the CloudFormation template of the application for the environment.
-func (opts *PackageAppOpts) Execute() error {
-	env, err := opts.store.GetEnvironment(opts.ProjectName(), opts.EnvName)
+func (o *PackageAppOpts) Execute() error {
+	env, err := o.store.GetEnvironment(o.ProjectName(), o.EnvName)
 	if err != nil {
 		return err
 	}
 
-	if opts.OutputDir != "" {
-		if err := opts.setFileWriters(); err != nil {
+	if o.OutputDir != "" {
+		if err := o.setFileWriters(); err != nil {
 			return err
 		}
 	}
 
-	templates, err := opts.getTemplates(env)
+	templates, err := o.getTemplates(env)
 	if err != nil {
 		return err
 	}
-	if _, err = opts.stackWriter.Write([]byte(templates.stack)); err != nil {
+	if _, err = o.stackWriter.Write([]byte(templates.stack)); err != nil {
 		return err
 	}
-	_, err = opts.paramsWriter.Write([]byte(templates.configuration))
+	_, err = o.paramsWriter.Write([]byte(templates.configuration))
 	return err
 }
 
-func (opts *PackageAppOpts) listAppNames() ([]string, error) {
-	apps, err := opts.ws.Apps()
+func (o *PackageAppOpts) askAppName() error {
+	if o.AppName != "" {
```

Comment:
Nit: feel like this should be placed out of this function, since `askAppName()` means asking for application name. And we probably should check if we need to ask for app name first.

Reply:
I see what you're saying. 
I wrote it this way to make `Ask()` easier to read, otherwise we end up with:
```go
if o.AppName != "" {
   if err := o.askAppName(); err != nil {
     return err
   }
}
// repeat for other flags
```
but then we can give the same argument for `Ask()` and bubble the if-statements back up :P

Is there a different name you'd prefer? We've been using this in other commands.

Reply:
Not a big one. I think it's up to you :) I am ok if you just keep as what it is now.